llmfoundry/eval/datasets/in_context_learning_evaluation.py

```diff
@@ -0,0 +1,1791 @@
+                chunked[k] = _default_split_batch(v, microbatch_size)
+            else:
+                raise ValueError(f'Unexpected key {k} in batch splitting')
+        num_chunks = len(chunked['input_ids'])
+        for k, v in batch.items():
+            if k in self.static_keys:
+                chunked[k] = [v] * num_chunks
+
+        batched_list = [
+            {k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)
+        ]
+        return batched_list
+
+
+class InContextLearningGenerationWithAnswersTaskDataset(InContextLearningDataset
+                                                       ):
+    """A dataset that constructs batches for in-context learning generation.
+
+    tasks with answers. Generation tasks evaluate a model's ability to
+    generate responses and score them against a set of gold-standard answers.
+
+    The input format is expected to be a jsonl file with the following fields:
+    - context: The question
+    - answer: The preferred answer to the question
+    - aliases: A list of aliases for the answer
+
+    See InContextLearningDataset for more details.
+
+    Additional Args:
+        cot_delimiter (str): Delimiter to place between the chain of thought and continuations.
+        early_stopping_criteria (Optional[List[str]]): Optional strings to trigger early stopping.
+        do_normalization (bool): Flag indicating whether to normalize generations before providing output.
+    """
+
+    def __init__(
+            self,
+            cot_delimiter: str = '',
+            early_stopping_criteria: Optional[List[str]] = None,
+            do_normalization: bool = True,
+            *args,  # pyright: ignore
+            **kwargs):  # pyright: ignore
+        if kwargs['tokenizer'].eos_token_id is None:
+            raise ValueError(
+                '`InContextLearningGenerationWithAnswersTaskDataset` tokenizer must have non-null `eos_token_id`'
+            )
+        self.cot_delimiter = cot_delimiter
+        self.has_cot = False
+        self.max_answer_length = 0
+        static_keys = [
+            'mode', 'cot_delimiter', 'generation_length', 'generation_kwargs',
+            'do_normalization', 'stopping_criteria'
+        ]
+        tensor_keys = ['input_ids', 'attention_mask']
+        list_keys = ['labels']
+        super().__init__(padding_side='left',
+                         tokenize_labels=False,
+                         static_keys=static_keys,
+                         list_keys=list_keys,
+                         tensor_keys=tensor_keys,
+                         *args,
+                         **kwargs)
+        # NOTE: set these after init call because they take class vars
+        self.early_stopping_criteria = early_stopping_criteria
+        self.base_batch = {
+            'input_ids': [],
+            'mode': 'generate',
+            'labels': [],
+            'cot_delimiter': self.cot_delimiter,
+            'generation_length': self.max_answer_length,
```

Comment:
we should make sure we port these changes: https://github.com/mosaicml/composer/pull/3014